From 5b6ec6f2f91926e9d1aa223f819f6567a606ce4f Mon Sep 17 00:00:00 2001
From: Xulin Sun <xulin.sun@windriver.com>
Date: Sat, 27 Jun 2020 17:20:52 +0800
Subject: [PATCH 1/2] driver: sdk_fman: replace local_irq_save with
 local_irq_save_nort

For preempt_rt kernel, need to replace local_irq_save/local_irq_restore with
local_irq_save_nort/local_irq_restore_nort, in order to fix below call trace:
[    2.604187] BUG: sleeping function called from invalid context at kernel/locking/rtmutex.c:975
[    2.604191] in_atomic(): 0, irqs_disabled(): 128, pid: 1, name: swapper/0
[    2.604194] CPU: 2 PID: 1 Comm: swapper/0 Not tainted 5.2.45-rt15-yocto-preempt-rt #1
[    2.604198] Hardware name: LS1043A RDB Board (DT)
[    2.604201] Call trace:
[    2.604202]  dump_backtrace+0x0/0x138
[    2.604210]  show_stack+0x24/0x30
[    2.604213]  dump_stack+0x9c/0xc4
[    2.604219]  ___might_sleep+0x13c/0x168
[    2.604223]  rt_spin_lock+0x44/0x80
[    2.604228]  get_page_from_freelist+0x454/0x1690
[    2.604233]  __alloc_pages_nodemask+0x158/0xff0
[    2.604237]  kmalloc_order+0x38/0x78
[    2.604241]  kmalloc_order_trace+0x3c/0x130
[    2.604244]  __kmalloc+0x33c/0x3b0
[    2.604249]  xx_Malloc+0x2c/0x60
[    2.604254]  XX_Malloc+0x20/0x30
[    2.604257]  ReadFmDevTreeNode+0x80/0x8b8
[    2.604260]  fm_probe+0x34/0x1358
[    2.604262]  platform_drv_probe+0x58/0xa8
[    2.604268]  really_probe+0xcc/0x290
[    2.604271]  driver_probe_device+0x5c/0xf0
[    2.604273]  device_driver_attach+0x74/0x80
[    2.604276]  __driver_attach+0x64/0xe0
[    2.604279]  bus_for_each_dev+0x80/0xd0
[    2.604282]  driver_attach+0x30/0x40
[    2.604285]  bus_add_driver+0x14c/0x1f0
[    2.604287]  driver_register+0x64/0x110
[    2.604290]  __platform_driver_register+0x54/0x60
[    2.604294]  LNXWRP_FM_Init+0x60/0x70
[    2.604296]  fm_load+0x14/0x58
[    2.604300]  do_one_initcall+0x64/0x2c8
[    2.604304]  kernel_init_freeable+0x338/0x3d8
[    2.604307]  kernel_init+0x18/0x104
[    2.604310]  ret_from_fork+0x10/0x1c

Signed-off-by: Xulin Sun <xulin.sun@windriver.com>
---
 .../net/ethernet/freescale/sdk_fman/src/xx/xx_arm_linux.c    | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/drivers/net/ethernet/freescale/sdk_fman/src/xx/xx_arm_linux.c b/drivers/net/ethernet/freescale/sdk_fman/src/xx/xx_arm_linux.c
index fe3661929909..fc726ece47a3 100644
--- a/drivers/net/ethernet/freescale/sdk_fman/src/xx/xx_arm_linux.c
+++ b/drivers/net/ethernet/freescale/sdk_fman/src/xx/xx_arm_linux.c
@@ -92,6 +92,11 @@
 
 #define __ERR_MODULE__      MODULE_UNKNOWN
 
+#ifdef CONFIG_PREEMPT_RT_FULL
+#define local_irq_save_nort
+#define local_irq_restore_nort
+#endif
+
 #ifdef BIGPHYSAREA_ENABLE
 #define MAX_ALLOCATION_SIZE     128 * 1024 /* Maximum size allocated with kmalloc is 128K */
 
-- 
2.17.1

